File: crates/suspect-codegen/src/dart_sdk/json.dart
/// Immutable JSON values. Numeric tokens never pass through `num` or `double`.
sealed class JsonValue {
  const JsonValue();

  /// Encode an exact JSON value under finite output and work budgets.
  String encode({JsonLimits limits = const JsonLimits()}) =>
      writeJson(this, limits: limits);
}

/// JSON null, distinct from a missing model property.
final class JsonNull extends JsonValue {
  const JsonNull();
}

/// An exact JSON boolean.
final class JsonBoolean extends JsonValue {
  const JsonBoolean(this.value);
  final bool value;
}

/// A Unicode-scalar JSON string. Unpaired UTF-16 surrogates are rejected.
final class JsonString extends JsonValue {
  JsonString(this.value) {
    _unicodeLength(value);
  }
  final String value;
}

/// A snapshot of a JSON array. The supplied collection is copied.
final class JsonArray extends JsonValue {
  JsonArray(Iterable<JsonValue> values) : values = List.unmodifiable(values);
  final List<JsonValue> values;
}

/// A snapshot of a JSON object, retaining all decoded member names.
final class JsonObject extends JsonValue {
  JsonObject(Map<String, JsonValue> values)
    : values = Map.unmodifiable(values) {
    for (final key in values.keys) {
      _unicodeLength(key);
    }
  }
  final Map<String, JsonValue> values;
}

/// A JSON numeric token with exact, symbolic-decimal semantics.
///
/// `1.0`, `-0.0` and `1e999999999999999999999` retain their original spellings.
/// Magnitude does not allocate expanded powers of ten. Use [compareTo] for
/// mathematical equality; object equality remains identity equality.
final class JsonNumber extends JsonValue implements Comparable<JsonNumber> {
  JsonNumber._(this.token, this._exact);

  /// Parse one complete JSON number, rejecting whitespace and non-JSON syntax.
  factory JsonNumber.parse(String token, {int maxBytes = 4096}) {
    _numberSyntax(token, maxBytes);
    return JsonNumber._(token, _Exact.parse(token));
  }

  /// Convert an already exact Dart integer without a floating-point step.
  factory JsonNumber.fromInt(int value) => JsonNumber.parse(value.toString());

  /// Convert an arbitrary-precision integer under the numeric token ceiling.
  factory JsonNumber.fromBigInt(BigInt value, {int maxBytes = 4096}) =>
      JsonNumber.parse(value.toString(), maxBytes: maxBytes);

  final String token;
  final _Exact _exact;

  /// Whether this value is a mathematical integer, independently of spelling.
  bool get isInteger => _exact.sign == 0 || _exact.exponent >= BigInt.zero;

  @override
  int compareTo(JsonNumber other) => _exact.compare(other._exact);

  /// Expand a mathematical integer only when its digit count fits [maxDigits].
  BigInt toBigInt({int maxDigits = 4096}) {
    if (maxDigits < 1 || maxDigits > 65536) {
      throw ArgumentError.value(maxDigits, 'maxDigits', 'must be 1..65536');
    }
    if (!isInteger) {
      throw const JsonException('not an exact integer');
    }
    if (_exact.sign == 0) {
      return BigInt.zero;
    }
    if (_exact.exponent + BigInt.from(_exact.digits.length) >
        BigInt.from(maxDigits)) {
      throw const JsonException(
        'integer expansion budget exhausted',
        resourceLimit: true,
      );
    }
    final coefficient = BigInt.parse(_exact.digits, radix: 10);
    final value = coefficient * BigInt.from(10).pow(_exact.exponent.toInt());
    return _exact.sign < 0 ? -value : value;
  }

  @override
  String toString() => token;
}

/// A mathematically integral JSON number, with its original wire token intact.
final class JsonInteger extends JsonNumber {
  JsonInteger._(super.token, super.exact) : super._();

  factory JsonInteger.parse(String token, {int maxBytes = 4096}) {
    final number = JsonNumber.parse(token, maxBytes: maxBytes);
    if (!number.isInteger) {
      throw const JsonException('not an exact integer');
    }
    return JsonInteger._(number.token, number._exact);
  }

  factory JsonInteger.fromInt(int value) => JsonInteger.parse(value.toString());
  factory JsonInteger.fromBigInt(BigInt value, {int maxBytes = 4096}) =>
      JsonInteger.parse(value.toString(), maxBytes: maxBytes);
}

/// Finite resource ceilings for one parser/writer invocation.
final class JsonLimits {
  const JsonLimits({
    this.maxBytes = 8 * 1024 * 1024,
    this.maxDepth = 128,
    this.maxSteps = 32 * 1024 * 1024,
    this.maxNumberBytes = 4096,
  });
  final int maxBytes;
  final int maxDepth;
  final int maxSteps;
  final int maxNumberBytes;

  void _check() {
    if (maxBytes < 0 ||
        maxBytes > 2147483647 ||
        maxDepth < 1 ||
        maxDepth > 128 ||
        maxSteps < 0 ||
        maxSteps > 2147483647 ||
        maxNumberBytes < 0 ||
        maxNumberBytes > 65536) {
      throw ArgumentError('invalid finite JSON resource policy');
    }
  }
}

/// Invalid JSON and finite-budget failures have distinct [resourceLimit] values.
final class JsonException implements Exception {
  const JsonException(this.message, {this.resourceLimit = false, this.offset});
  final String message;
  final bool resourceLimit;
  final int? offset;
  @override
  String toString() => 'JsonException: $message';
}

/// Parse exact JSON. Duplicate decoded keys and invalid Unicode are rejected.
JsonValue parseJson(String text, {JsonLimits limits = const JsonLimits()}) {
  limits._check();
  if (text.length > limits.maxBytes || _unicodeLength(text) > limits.maxBytes) {
    throw const JsonException(
      'JSON input byte budget exhausted',
      resourceLimit: true,
    );
  }
  return _JsonParser(text, limits).parse();
}

/// Parse strict UTF-8 JSON, checking byte length before decoding or allocation.
JsonValue parseJsonBytes(
  List<int> bytes, {
  JsonLimits limits = const JsonLimits(),
}) {
  limits._check();
  if (bytes.length > limits.maxBytes) {
    throw const JsonException(
      'JSON input byte budget exhausted',
      resourceLimit: true,
    );
  }
  try {
    // Invalid integers in an injected byte list must not be truncated to bytes.
    for (final byte in bytes) {
      if (byte < 0 || byte > 255) {
        throw const FormatException();
      }
    }
    return parseJson(utf8.decode(bytes, allowMalformed: false), limits: limits);
  } on FormatException {
    throw const JsonException('invalid UTF-8 JSON input');
  }
}

/// Serialize tokens verbatim, preserving omission decisions made by codecs.
String writeJson(JsonValue value, {JsonLimits limits = const JsonLimits()}) {
  limits._check();
  final writer = _JsonWriter(limits);
  writer.value(value, 0);
  return writer.out.toString();
}

int _unicodeLength(String value) {
  var bytes = 0;
  for (var i = 0; i < value.length; i++) {
    final unit = value.codeUnitAt(i);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      if (++i >= value.length) {
        throw const JsonException('unpaired Unicode surrogate');
      }
      final next = value.codeUnitAt(i);
      if (next < 0xdc00 || next > 0xdfff) {
        throw const JsonException('unpaired Unicode surrogate');
      }
      bytes += 4;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      throw const JsonException('unpaired Unicode surrogate');
    } else {
      bytes += unit < 0x80
          ? 1
          : unit < 0x800
          ? 2
          : 3;
    }
  }
  return bytes;
}

bool _digit(int unit) => unit >= 48 && unit <= 57;

void _numberSyntax(String text, int maxBytes) {
  if (maxBytes < 0 || maxBytes > 65536) {
    throw ArgumentError.value(maxBytes, 'maxBytes');
  }
  if (text.length > maxBytes) {
    throw const JsonException(
      'numeric token budget exhausted',
      resourceLimit: true,
    );
  }
  var i = 0;
  if (i < text.length && text.codeUnitAt(i) == 45) {
    i++;
  }
  if (i >= text.length) {
    throw const JsonException('invalid JSON number');
  }
  if (text.codeUnitAt(i) == 48) {
    i++;
  } else if (text.codeUnitAt(i) >= 49 && text.codeUnitAt(i) <= 57) {
    while (i < text.length && _digit(text.codeUnitAt(i))) {
      i++;
    }
  } else {
    throw const JsonException('invalid JSON number');
  }
  if (i < text.length && text.codeUnitAt(i) == 46) {
    final start = ++i;
    while (i < text.length && _digit(text.codeUnitAt(i))) {
      i++;
    }
    if (i == start) {
      throw const JsonException('invalid JSON fraction');
    }
  }
  if (i < text.length &&
      (text.codeUnitAt(i) == 69 || text.codeUnitAt(i) == 101)) {
    i++;
    if (i < text.length &&
        (text.codeUnitAt(i) == 43 || text.codeUnitAt(i) == 45)) {
      i++;
    }
    final start = i;
    while (i < text.length && _digit(text.codeUnitAt(i))) {
      i++;
    }
    if (i == start) {
      throw const JsonException('invalid JSON exponent');
    }
  }
  if (i != text.length) {
    throw const JsonException('invalid JSON number');
  }
}

final class _Exact {
  _Exact(this.sign, this.digits, this.exponent);
  final int sign;
  final String digits;
  final BigInt exponent;

  factory _Exact.parse(String token) {
    final negative = token.startsWith('-');
    final unsigned = negative ? token.substring(1) : token;
    final e = unsigned.indexOf(RegExp('[eE]'));
    final mantissa = e < 0 ? unsigned : unsigned.substring(0, e);
    final dot = mantissa.indexOf('.');
    final fraction = dot < 0 ? 0 : mantissa.length - dot - 1;
    final digits = mantissa.replaceFirst('.', '');
    var start = 0;
    while (start < digits.length && digits.codeUnitAt(start) == 48) {
      start++;
    }
    if (start == digits.length) {
      return _Exact(0, '0', BigInt.zero);
    }
    var end = digits.length;
    while (digits.codeUnitAt(end - 1) == 48) {
      end--;
    }
    final exponent = e < 0
        ? BigInt.zero
        : BigInt.parse(unsigned.substring(e + 1), radix: 10);
    return _Exact(
      negative ? -1 : 1,
      digits.substring(start, end),
      exponent - BigInt.from(fraction) + BigInt.from(digits.length - end),
    );
  }

  int compare(_Exact other) {
    if (sign != other.sign) {
      return sign.compareTo(other.sign);
    }
    if (sign == 0) {
      return 0;
    }
    final order = (exponent + BigInt.from(digits.length)).compareTo(
      other.exponent + BigInt.from(other.digits.length),
    );
    if (order != 0) {
      return sign * order;
    }
    final length = digits.length > other.digits.length
        ? digits.length
        : other.digits.length;
    for (var i = 0; i < length; i++) {
      final a = i < digits.length ? digits.codeUnitAt(i) : 48;
      final b = i < other.digits.length ? other.digits.codeUnitAt(i) : 48;
      if (a != b) {
        return sign * a.compareTo(b);
      }
    }
    return 0;
  }

  bool multipleOf(_Exact divisor, void Function(int) spend) {
    if (sign == 0) {
      return true;
    }
    final shift = exponent - divisor.exponent;
    if (shift < BigInt.zero) {
      return false;
    }
    spend(
      digits.length +
          divisor.digits.length +
          digits.length * divisor.digits.length,
    );
    final numerator = BigInt.parse(digits, radix: 10);
    var denominator = BigInt.parse(divisor.digits, radix: 10);
    // Remove only the factors supplied by symbolic 10^shift. Loop work is
    // bounded by the written coefficient, never the exponent's magnitude.
    for (final prime in [BigInt.two, BigInt.from(5)]) {
      var powers = BigInt.zero;
      while (powers < shift && denominator % prime == BigInt.zero) {
        spend(divisor.digits.length);
        denominator ~/= prime;
        powers += BigInt.one;
      }
    }
    return numerator % denominator == BigInt.zero;
  }
}

final class _JsonParser {
  _JsonParser(this.text, this.limits) : left = limits.maxSteps;
  final String text;
  final JsonLimits limits;
  int at = 0;
  int left;
  void spend([int n = 1]) {
    left -= n;
    if (left < 0) {
      fail('JSON parser work budget exhausted', limit: true);
    }
  }

  Never fail(String message, {bool limit = false}) =>
      throw JsonException(message, resourceLimit: limit, offset: at);
  int get peek => at < text.length ? text.codeUnitAt(at) : -1;
  void whitespace() {
    while (peek == 32 || peek == 9 || peek == 10 || peek == 13) {
      spend();
      at++;
    }
  }

  JsonValue parse() {
    whitespace();
    final result = value(0);
    whitespace();
    if (at != text.length) {
      fail('trailing JSON input');
    }
    return result;
  }

  JsonValue value(int depth) {
    spend();
    if (depth >= limits.maxDepth) {
      fail('JSON nesting budget exhausted', limit: true);
    }
    switch (peek) {
      case 110:
        literal('null');
        return const JsonNull();
      case 116:
        literal('true');
        return const JsonBoolean(true);
      case 102:
        literal('false');
        return const JsonBoolean(false);
      case 34:
        return JsonString(string());
      case 91:
        at++;
        whitespace();
        final values = <JsonValue>[];
        if (peek == 93) {
          at++;
          return JsonArray(values);
        }
        while (true) {
          values.add(value(depth + 1));
          whitespace();
          if (peek == 93) {
            at++;
            return JsonArray(values);
          }
          if (peek != 44) {
            fail('expected array comma');
          }
          at++;
          whitespace();
        }
      case 123:
        at++;
        whitespace();
        final values = <String, JsonValue>{};
        if (peek == 125) {
          at++;
          return JsonObject(values);
        }
        while (true) {
          if (peek != 34) {
            fail('expected object key');
          }
          final key = string();
          whitespace();
          if (values.containsKey(key)) {
            fail('duplicate decoded object key');
          }
          if (peek != 58) {
            fail('expected object colon');
          }
          at++;
          whitespace();
          values[key] = value(depth + 1);
          whitespace();
          if (peek == 125) {
            at++;
            return JsonObject(values);
          }
          if (peek != 44) {
            fail('expected object comma');
          }
          at++;
          whitespace();
        }
      default:
        final start = at;
        while (_digit(peek) ||
            peek == 45 ||
            peek == 43 ||
            peek == 46 ||
            peek == 101 ||
            peek == 69) {
          spend();
          at++;
          if (at - start > limits.maxNumberBytes) {
            fail('numeric token budget exhausted', limit: true);
          }
        }
        if (at == start) {
          fail('expected JSON value');
        }
        return JsonNumber.parse(
          text.substring(start, at),
          maxBytes: limits.maxNumberBytes,
        );
    }
  }

  void literal(String expected) {
    spend(expected.length);
    if (!text.startsWith(expected, at)) {
      fail('invalid JSON literal');
    }
    at += expected.length;
  }

  int hex() {
    var result = 0;
    for (var i = 0; i < 4; i++) {
      spend();
      final c = peek;
      final digit = c >= 48 && c <= 57
          ? c - 48
          : c >= 65 && c <= 70
          ? c - 55
          : c >= 97 && c <= 102
          ? c - 87
          : -1;
      if (digit < 0) {
        fail('invalid Unicode escape');
      }
      at++;
      result = result * 16 + digit;
    }
    return result;
  }

  String string() {
    at++;
    final result = StringBuffer();
    while (at < text.length) {
      spend();
      var unit = text.codeUnitAt(at++);
      if (unit == 34) {
        return result.toString();
      }
      if (unit < 32) {
        fail('unescaped control character');
      }
      if (unit == 92) {
        if (at >= text.length) {
          fail('incomplete string escape');
        }
        spend();
        switch (text.codeUnitAt(at++)) {
          case 34:
            unit = 34;
          case 92:
            unit = 92;
          case 47:
            unit = 47;
          case 98:
            unit = 8;
          case 102:
            unit = 12;
          case 110:
            unit = 10;
          case 114:
            unit = 13;
          case 116:
            unit = 9;
          case 117:
            unit = hex();
            if (unit >= 0xd800 && unit <= 0xdbff) {
              if (!text.startsWith('\\u', at)) {
                fail('unpaired Unicode surrogate');
              }
              at += 2;
              final next = hex();
              if (next < 0xdc00 || next > 0xdfff) {
                fail('unpaired Unicode surrogate');
              }
              result.writeCharCode(unit);
              result.writeCharCode(next);
              continue;
            }
            if (unit >= 0xdc00 && unit <= 0xdfff) {
              fail('unpaired Unicode surrogate');
            }
          default:
            fail('invalid string escape');
        }
      }
      result.writeCharCode(unit);
    }
    fail('unterminated JSON string');
  }
}

final class _JsonWriter {
  _JsonWriter(this.limits) : left = limits.maxSteps;
  final JsonLimits limits;
  final StringBuffer out = StringBuffer();
  int bytes = 0;
  int left;
  void spend([int n = 1]) {
    left -= n;
    if (left < 0) {
      throw const JsonException(
        'JSON writer work budget exhausted',
        resourceLimit: true,
      );
    }
  }

  void add(String text) {
    spend(text.length);
    bytes += _unicodeLength(text);
    if (bytes > limits.maxBytes) {
      throw const JsonException(
        'JSON output byte budget exhausted',
        resourceLimit: true,
      );
    }
    out.write(text);
  }

  void string(String text) {
    spend(text.length);
    if (text.length > limits.maxBytes - bytes) {
      throw const JsonException(
        'JSON output byte budget exhausted',
        resourceLimit: true,
      );
    }
    _unicodeLength(text);
    add('"');
    // Write bounded slices; escaped strings are never built in a separate
    // unbounded buffer before checking the output ceiling.
    for (var i = 0; i < text.length; i++) {
      final unit = text.codeUnitAt(i);
      if (unit == 34) {
        add('\\"');
      } else if (unit == 92) {
        add('\\\\');
      } else if (unit < 32) {
        add('\\u${unit.toRadixString(16).padLeft(4, '0')}');
      } else if (unit >= 0xd800 && unit <= 0xdbff) {
        add(text.substring(i, i + 2));
        i++;
      } else {
        add(text[i]);
      }
    }
    add('"');
  }

  void value(JsonValue value, int depth) {
    spend();
    if (depth >= limits.maxDepth) {
      throw const JsonException(
        'JSON nesting budget exhausted',
        resourceLimit: true,
      );
    }
    switch (value) {
      case JsonNull():
        add('null');
      case JsonBoolean():
        add(value.value ? 'true' : 'false');
      case JsonString():
        string(value.value);
      case JsonNumber():
        if (value.token.length > limits.maxNumberBytes) {
          throw const JsonException(
            'numeric token budget exhausted',
            resourceLimit: true,
          );
        }
        add(value.token);
      case JsonArray():
        add('[');
        for (var i = 0; i < value.values.length; i++) {
          if (i != 0) {
            add(',');
          }
          this.value(value.values[i], depth + 1);
        }
        add(']');
      case JsonObject():
        add('{');
        var first = true;
        for (final entry in value.values.entries) {
          if (!first) {
            add(',');
          }
          first = false;
          string(entry.key);
          add(':');
          this.value(entry.value, depth + 1);
        }
        add('}');
    }
  }
}

File: crates/suspect-codegen/src/ruby_sdk/json.rb
# frozen_string_literal: true

module __NAMESPACE__
  # Base for SDK-owned failures. Messages never contain response bodies or credentials.
  class SdkError < StandardError
    attr_reader :kind, :source, :instance_path, :operation_id, :status, :headers, :capture, :truncated

    def initialize(message, kind: :sdk, source: nil, instance_path: '', operation_id: nil,
                   status: nil, headers: {}, capture: ''.b, truncated: false)
      super(message)
      @kind, @source, @instance_path, @operation_id = kind, source, instance_path, operation_id
      @status, @headers, @capture, @truncated = status, headers, capture, truncated
    end
  end

  # Invalid JSON syntax, native JSON-domain value, or JSON resource exhaustion.
  class JsonError < SdkError
    def initialize(message, **context)
      super(message, kind: :json, **context)
    end
  end

  # A value cannot be represented by the source-bound native model.
  class CodecError < SdkError
    def initialize(message, kind: :codec, **context)
      super(message, kind: kind, **context)
    end
  end

  # Completed schema rejection; distinct from incomplete evaluation.
  class ValidationError < CodecError
    def initialize(message, **context)
      super(message, kind: :invalid, **context)
    end
  end

  # Resource exhaustion or nonproductive recursion. Logical trials cannot suppress it.
  class EvaluationFailure < CodecError
    def initialize(message, **context)
      super(message, kind: :evaluation_failure, **context)
    end
  end

  # Explicit absence. Only {UNSET} is an absent value; nil is JSON null.
  class Unset
    def inspect = 'UNSET'
    alias to_s inspect
  end
  UNSET = Unset.new.freeze
  Unset.private_class_method(:new)

  # Exact JSON numeric token, including decimals and symbolic unbounded exponents.
  # Integer inputs are also accepted by numeric codecs. Float is never coerced.
  # @example Exact construction
  #   JsonNumber.new('9007199254740993.000000000000000001').token
  class JsonNumber
    include Comparable
    TOKEN = /\A-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?\z/n.freeze
    attr_reader :token

    def initialize(token)
      unless token.instance_of?(String) && token.bytesize <= Internal::POLICY[:max_number_bytes] && token.ascii_only? && TOKEN.match?(token)
        raise JsonError, 'expected a bounded exact JSON numeric token'
      end
      @token = token.dup.freeze
      freeze
    end

    def <=>(other)
      return nil unless other.instance_of?(JsonNumber) || other.instance_of?(Integer)
      Internal::Exact.new(@token).compare(Internal::Exact.new(Internal.number_token(other)))
    end

    def eql?(other) = other.instance_of?(JsonNumber) && (self <=> other).zero?
    def hash = Internal::Exact.new(@token).key.hash
    def integer? = Internal::Exact.new(@token).integral?
    def inspect = "JsonNumber(#{@token.inspect})"
    alias to_s token

    # Explicit bounded conversion. Huge exponents are never expanded implicitly.
    # @param max_digits [Integer] maximum expanded decimal digits, at most 4096
    # @return [Integer]
    # @raise [RangeError] non-integral value or expansion outside the requested bound
    def to_i(max_digits: 4096)
      unless max_digits.instance_of?(Integer) && max_digits.between?(1, 4096)
        raise ArgumentError, 'max_digits must be in 1..4096'
      end
      number = Internal::Exact.new(@token)
      raise RangeError, 'JSON number is not mathematically integral' unless number.integral?
      return 0 if number.sign.zero?
      raise RangeError, 'integer expansion exceeds max_digits' if number.digits.length + number.exponent > max_digits
      number.sign * number.digits.to_i * (10**number.exponent)
    end

    # Explicit lossy conversion, never called by a codec or transport.
    # @return [Float]
    def to_f
      value = Float(@token)
      raise RangeError, 'number is outside finite Float range' unless value.finite?
      value
    end
  end

  # Strict exact JSON. Duplicate keys, lone surrogates, non-finite numbers,
  # object hooks, symbols, cyclic containers and binary floats are rejected.
  module Json
    module_function

    # @return [nil, Boolean, String, JsonNumber, Array, Hash]
    def parse(text, max_bytes: Internal::POLICY[:max_response_bytes],
              max_depth: Internal::POLICY[:max_json_depth], max_work: Internal::POLICY[:max_json_work])
      Reader.new(text, max_bytes, max_depth, max_work).read
    end

    # @return [String] UTF-8 JSON preserving every numeric token
    def dump(value, max_bytes: Internal::POLICY[:max_request_bytes],
             max_depth: Internal::POLICY[:max_json_depth], max_work: Internal::POLICY[:max_json_work])
      Writer.new(max_bytes, max_depth, max_work).write(value)
    end

    # @api private
    class Limits
      def initialize(max_bytes, max_depth, max_work)
        unless [max_bytes, max_work].all? { |n| n.instance_of?(Integer) && n.between?(1, 128 * 1024 * 1024) } &&
               max_depth.instance_of?(Integer) && max_depth.between?(1, 128)
          raise ArgumentError, 'JSON limits must be positive and within the finite runtime profile'
        end
        @max_bytes, @max_depth, @work = max_bytes, max_depth, max_work
      end

      def spend(amount = 1)
        @work -= amount
        raise JsonError, 'JSON work limit exceeded' if @work.negative?
      end

      def depth!(depth)
        raise JsonError, 'JSON nesting limit exceeded' if depth > @max_depth
      end
    end

    # @api private
    class Reader < Limits
      NUMBER = /\G-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/n.freeze
      HEX = /\A[0-9a-fA-F]{4}\z/n.freeze

      def initialize(text, max_bytes, max_depth, max_work)
        super(max_bytes, max_depth, max_work)
        raise JsonError, 'JSON input must be a String' unless text.instance_of?(String)
        raise JsonError, 'JSON input byte limit exceeded' if text.bytesize > @max_bytes
        spend(text.bytesize)
        @text = text.dup.force_encoding(Encoding::UTF_8)
        raise JsonError, 'JSON input is not valid UTF-8' unless @text.valid_encoding?
        @text.force_encoding(Encoding::BINARY)
        @at = 0
      end

      def read
        value = value(0)
        whitespace
        fail_json('trailing JSON bytes') unless @at == @text.bytesize
        value
      end

      private

      def fail_json(message)
        raise JsonError, "#{message} at byte #{@at}"
      end

      def whitespace
        while [9, 10, 13, 32].include?(@text.getbyte(@at))
          spend
          @at += 1
        end
      end

      def value(depth)
        depth!(depth)
        spend
        whitespace
        case @text.getbyte(@at)
        when 34 then string
        when 123 then object(depth)
        when 91 then array(depth)
        when 116 then literal('true', true)
        when 102 then literal('false', false)
        when 110 then literal('null', nil)
        else
          match = NUMBER.match(@text, @at)
          fail_json('expected a JSON value') unless match
          token = match[0]
          @at = match.end(0)
          JsonNumber.new(token)
        end
      end

      def literal(token, value)
        fail_json('invalid JSON literal') unless @text.byteslice(@at, token.bytesize) == token
        @at += token.bytesize
        value
      end

      def object(depth)
        @at += 1
        result = {}
        whitespace
        if @text.getbyte(@at) == 125
          @at += 1
          return result
        end
        loop do
          spend
          whitespace
          fail_json('expected an object key') unless @text.getbyte(@at) == 34
          key = string
          fail_json('duplicate decoded object key') if result.key?(key)
          whitespace
          fail_json('expected colon') unless @text.getbyte(@at) == 58
          @at += 1
          result[key] = value(depth + 1)
          whitespace
          byte = @text.getbyte(@at)
          @at += 1
          break if byte == 125
          fail_json('expected comma or object end') unless byte == 44
        end
        result
      end

      def array(depth)
        @at += 1
        result = []
        whitespace
        if @text.getbyte(@at) == 93
          @at += 1
          return result
        end
        loop do
          result << value(depth + 1)
          whitespace
          byte = @text.getbyte(@at)
          @at += 1
          break if byte == 93
          fail_json('expected comma or array end') unless byte == 44
        end
        result
      end

      def hex_unit
        text = @text.byteslice(@at, 4)
        fail_json('invalid Unicode escape') unless text && HEX.match?(text)
        @at += 4
        text.to_i(16)
      end

      def string
        @at += 1
        result = +''.b
        loop do
          spend
          byte = @text.getbyte(@at)
          fail_json('unterminated JSON string') unless byte
          @at += 1
          break if byte == 34
          if byte == 92
            escape = @text.getbyte(@at)
            @at += 1
            case escape
            when 34, 47, 92 then result << escape
            when 98 then result << 8
            when 102 then result << 12
            when 110 then result << 10
            when 114 then result << 13
            when 116 then result << 9
            when 117
              unit = hex_unit
              if unit.between?(0xD800, 0xDBFF)
                fail_json('missing low surrogate') unless @text.byteslice(@at, 2) == '\\u'
                @at += 2
                low = hex_unit
                fail_json('invalid low surrogate') unless low.between?(0xDC00, 0xDFFF)
                unit = 0x10000 + ((unit - 0xD800) << 10) + low - 0xDC00
              elsif unit.between?(0xDC00, 0xDFFF)
                fail_json('unpaired low surrogate')
              end
              result << [unit].pack('U').b
            else fail_json('invalid string escape')
            end
          elsif byte < 32
            fail_json('unescaped string control character')
          else
            result << byte
          end
        end
        result.force_encoding(Encoding::UTF_8)
        fail_json('invalid UTF-8 string') unless result.valid_encoding?
        result
      end
    end

    # @api private
    class Writer < Limits
      def initialize(max_bytes, max_depth, max_work)
        super
        @output, @active = +'', {}
      end

      def write(value)
        value(value, 0)
        @output
      end

      private

      def append(text)
        raise JsonError, 'JSON output byte limit exceeded' if @output.bytesize + text.bytesize > @max_bytes
        spend(text.bytesize)
        @output << text
      end

      def string(text)
        Internal.utf8!(text)
        spend(text.bytesize)
        append('"')
        text.each_codepoint do |codepoint|
          case codepoint
          when 34 then append('\\"')
          when 92 then append('\\\\')
          when 0...32 then append(format('\\u%04x', codepoint))
          else append([codepoint].pack('U'))
          end
        end
        append('"')
      end

      def value(value, depth)
        depth!(depth)
        spend
        case
        when value.nil? then append('null')
        when value.equal?(true) then append('true')
        when value.equal?(false) then append('false')
        when value.instance_of?(String) then string(value)
        when value.instance_of?(JsonNumber), value.instance_of?(Integer) then append(Internal.number_token(value))
        when value.instance_of?(Array), value.instance_of?(Hash)
          identity = value.object_id
          raise JsonError, 'cyclic JSON container' if @active[identity]
          @active[identity] = true
          begin
            if value.instance_of?(Array)
              append('[')
              value.each_with_index { |child, i| append(',') unless i.zero?; value(child, depth + 1) }
              append(']')
            else
              value.each_key { |key| Internal.utf8!(key); spend(key.bytesize + 1) }
              append('{')
              value.keys.sort.each_with_index do |key, i|
                append(',') unless i.zero?
                string(key)
                append(':')
                value(value[key], depth + 1)
              end
              append('}')
            end
          ensure
            @active.delete(identity)
          end
        else raise JsonError, 'value is outside the exact JSON domain (Float and UNSET are not JSON values)'
        end
      end
    end
  end

  # @api private
  module Internal
    module_function

    def number_token(value)
      return value.token if value.instance_of?(JsonNumber)
      unless value.instance_of?(Integer) && value.bit_length <= POLICY[:max_number_bytes] * 4
        raise JsonError, 'expected a bounded exact numeric value'
      end
      token = value.to_s
      raise JsonError, 'numeric token byte limit exceeded' if token.bytesize > POLICY[:max_number_bytes]
      token
    end

    def utf8!(value)
      unless value.instance_of?(String) && value.valid_encoding? &&
             (value.encoding == Encoding::UTF_8 || value.ascii_only? && value.encoding.ascii_compatible?)
        raise JsonError, 'expected a UTF-8 string'
      end
      value
    end

    def freeze_tree(value)
      case value
      when Hash then value.each { |key, child| key.freeze; freeze_tree(child) }
      when Array then value.each { |child| freeze_tree(child) }
      end
      value.freeze
    end

    # Normalized exact decimal. Exponent magnitude is never expanded.
    class Exact
      attr_reader :sign, :digits, :exponent

      def initialize(token)
        negative = token.start_with?('-')
        mantissa, exponent = token.delete_prefix('-').split(/[eE]/, 2)
        whole, fraction = mantissa.split('.', 2)
        fraction ||= ''
        coefficient = (whole + fraction).sub(/\A0+/, '')
        if coefficient.empty?
          @sign, @digits, @exponent = 0, '0', 0
        else
          @sign = negative ? -1 : 1
          @digits = coefficient.sub(/0+\z/, '')
          @exponent = (exponent || '0').to_i - fraction.length + coefficient.length - @digits.length
        end
      end

      def key = [@sign, @digits, @exponent]
      def integral? = @sign.zero? || @exponent >= 0

      def compare(other)
        return @sign <=> other.sign if @sign != other.sign
        return 0 if @sign.zero?
        order = (@digits.length + @exponent) <=> (other.digits.length + other.exponent)
        if order.zero?
          length = [@digits.length, other.digits.length].max
          order = @digits.ljust(length, '0') <=> other.digits.ljust(length, '0')
        end
        @sign * order
      end

      def multiple_of?(divisor)
        return true if @sign.zero?
        shift = @exponent - divisor.exponent
        return false if shift.negative?
        yield(@digits.length + divisor.digits.length)
        a, b = @digits.to_i, divisor.digits.to_i
        [2, 5].each do |prime|
          powers = 0
          loop do
            yield([1, b.bit_length / 8].max)
            break if powers >= shift || (b % prime) != 0
            b /= prime
            powers += 1
          end
        end
        yield([1, a.bit_length / 8].max * [1, b.bit_length / 8].max)
        (a % b).zero?
      end
    end
  end
end
